### PR TITLE
Homework registration validation

### PR DESCRIPTION
**Pull request by Cosmic-M**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 1]
```diff
@@ -5,10 +5,51 @@
...
+            throw new UserPasswordNullException(PASSWORD_IS_NULL);
+        }
+        if (user.getPassword().replaceAll(SPECIAL_SYMBOL, "")
+                .length() - user.getPassword().length() != 0) {
+            throw new RuntimeException(PASSWORD_HAS_INVALID_SYMBOL);
+        }
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new RuntimeException(LOGIN_EXISTS);
+        }
+        return storageDao.add(user);
```
Comment (Ivan-Yuriv) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:53]: You can separate all these conditionals to different methods, but it is up to you.

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 2]
```diff
@@ -1,8 +1,118 @@
...
+    @Test
+    void userPasswordIsNull_NotOk() {
+        user.setPassword(null);
+        assertThrows(UserPasswordNullException.class,
+                () -> registrationService.register(user),
+                "User's password might not be null\n");
+    }
+
+    @Test
+    void userPasswordHasInvalidSymbols_NotOk() {
```
Comment (Ivan-Yuriv) [inline src/test/java/core/basesyntax/HelloWorldTest.java:102]: register_nullPassword_notOk

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 3]
```diff
@@ -1,8 +1,118 @@
...
+    @Test
+    void userLoginLesserThanSixCharacters_NotOk() {
+        user.setLogin("Robin");
+        assertThrows(RuntimeException.class,
+                () -> registrationService.register(user),
+                "too short login\n");
+    }
+
+    @Test
+    void userLoginIsNull_NotOk() {
```
Comment (Ivan-Yuriv) [inline src/test/java/core/basesyntax/HelloWorldTest.java:69]: void register_nullLogin_notOk()

you have to change naming of your Tests. 
In description to this task we have an example, how test has to be called. For example:
"register_nullAge_notOk" after underscope you have to write the next word starts from small letter.

**Review by Ivan-Yuriv — APPROVED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 4]
```diff
@@ -5,10 +5,51 @@
...
+    private static final String LOGIN_IS_NULL = "user's login is null";
+    private static final String USER_IS_NULL = "user is null";
+    private static final String PASSWORD_IS_NULL = "user's password is null";
+    private static final String TOO_YOUNG = "user must first grow up a little; age less than 18";
+    private static final String NEGATIVE_AGE = "negative user's age";
+    private static final String LOGIN_HAS_INVALID_SYMBOL = "login has at least one invalid symbol";
+    private static final String PASSWORD_HAS_INVALID_SYMBOL
+            = "Password has at least one invalid symbol";
+    private static final String LOGIN_HAS_LESS_THAT_SIX_SYMBOLS
+            = "login too short, must be at least six symbols";
```
Comment (KostyaNikiforov) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:18]: It's actually redundant to use constants for exception messages.
It just gets harder to work with it. You have to scroll to the top of the class to see the message while reading the code and it's almost impossible for you to reuse it.

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 5]
```diff
@@ -5,10 +5,51 @@
...
+        if (user.getPassword() == null) {
+            throw new UserPasswordNullException(PASSWORD_IS_NULL);
+        }
+        if (user.getPassword().replaceAll(SPECIAL_SYMBOL, "")
+                .length() - user.getPassword().length() != 0) {
+            throw new RuntimeException(PASSWORD_HAS_INVALID_SYMBOL);
+        }
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new RuntimeException(LOGIN_EXISTS);
+        }
```
Comment (KostyaNikiforov) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:52]: It is a very large method. Let's improve the readability by using private methods

File: src/main/java/core/basesyntax/service/UserHasNoLoginException.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.service;
+
+public class UserHasNoLoginException extends RuntimeException {
+
```
Comment (KostyaNikiforov) [inline src/main/java/core/basesyntax/service/UserHasNoLoginException.java:4]: Empty line

File: src/main/java/core/basesyntax/service/UserNullException.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.service;
+
+public class UserNullException extends RuntimeException {
+
```
Comment (KostyaNikiforov) [inline src/main/java/core/basesyntax/service/UserNullException.java:4]: And here

File: src/main/java/core/basesyntax/service/UserPasswordNullException.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.service;
+
+public class UserPasswordNullException extends RuntimeException {
+
```
Comment (KostyaNikiforov) [inline src/main/java/core/basesyntax/service/UserPasswordNullException.java:4]: And here

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 9]
```diff
@@ -1,8 +1,118 @@
...
+    void userNull_notOk() {
+        User userNull = null;
+        assertThrows(UserNullException.class,
+                () -> registrationService.register(userNull),
+                "user is null\n");
+    }
+
+    @Test
+    void userYoungerThan18_notOk() {
+        user.setAge(12);
```
Comment (KostyaNikiforov) [inline src/test/java/core/basesyntax/HelloWorldTest.java:46]: Let's take the maximum invalid value.
In this case, it's 17

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 10]
```diff
@@ -1,8 +1,118 @@
...
+        assertThrows(RuntimeException.class,
+                () -> registrationService.register(user),
+                "condition break: user with age up to 18 was added\n");
+    }
+
+    @Test
+    void userAgeNegative_notOk() {
+        user.setAge(-25);
```
Comment (KostyaNikiforov) [inline src/test/java/core/basesyntax/HelloWorldTest.java:54]: The same. Why did you take 25 here? Let's take the first negative number

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 11]
```diff
@@ -1,8 +1,118 @@
...
+    @BeforeEach
+     void createValidUser() {
+        user = new User();
+        user.setAge(42);
+        user.setLogin("labudabudapta");
+        user.setPassword("motorama");
+    }
+
+    @Test
+    void userNull_notOk() {
```
Comment (KostyaNikiforov) [inline src/test/java/core/basesyntax/HelloWorldTest.java:37]: Let's make the name of each test method as it was suggested in the readme.
There should be three parts:
First - the name of the testing method
Second - test case description
Third - expected behavior 
In this case it should look this way:
`register_nullUser_noOk`

**Review by KostyaNikiforov — CHANGES_REQUESTED** [review]
Also, you should add some positive cases, and what about the situation when the user already exists in the base?

**Comment by Cosmic-M** [comment]
Also, you should add some positive cases, and what about the situation when the user already exists in the base?
register_userWithDuplicateLoginMayBeAddToDatabase_notOk

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 12]
```diff
@@ -5,10 +5,53 @@
...
+
+    private void passwordValidate(String password) {
+        if (password == null) {
+            throw new UserPasswordNullException("user's password is null");
+        }
+        if (password.length() < LENGTH_OF_PASSWORD) {
+            throw new RuntimeException("password too short, must be at least six symbols");
+        }
+        if (password.replaceAll("\\W+", "")
+                .length() - password.length() != 0) {
```
Comment (KostyaNikiforov) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:53]: you can use the method `isBlank()`

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 13]
```diff
@@ -1,8 +1,166 @@
...
+    @BeforeEach
+     void createValidUser() {
+        user = new User();
+        user.setAge(42);
+        user.setLogin("labudabudapta");
+        user.setPassword("motorama");
+    }
+
+    @Test
+    void register_addSomeUsersWithUniqueLoginsToDatabase_Ok() {
```
Comment (KostyaNikiforov) [inline src/test/java/core/basesyntax/HelloWorldTest.java:38]:
````suggestion
    void register_addSomeUsersWithUniqueLoginsToDatabase_ok() {
````
Each part must start with a lowercase letter

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 14]
```diff
@@ -1,8 +1,166 @@
...
+        registrationService.register(user);
+        registrationService.register(user2);
+        assertEquals(2, Storage.people.size(),
+                "After adding two users with the same passwords but different "
+                        + "logins to database, the size must be 2, but actual size is "
+                        + Storage.people.size());
+    }
+
+    @Test
+    void addUserWhichLoginIncludesDigits_Ok() {
```
Comment (KostyaNikiforov) [inline src/test/java/core/basesyntax/HelloWorldTest.java:70]: naming

File: src/test/java/core/basesyntax/HelloWorldTest.java [hunk 15]
```diff
@@ -1,8 +1,166 @@
...
+
+    @AfterEach
+    void clearStorage() {
+        Storage.people.clear();
+    }
 
+    @AfterAll
+    static void clearingStorageWithTestingObjects() {
+        Storage.people.clear();
+    }
```
Comment (KostyaNikiforov) [inline src/test/java/core/basesyntax/HelloWorldTest.java:165]: Do we need this method, if we have AfterEach?

**Pull request closed without merging**